solidity/contracts/liquidity-protection/TokenTimeWeightedAverage.sol

```diff
@@ -0,0 +1,375 @@
+     *
+     * @param _token the token the data is accumulated for
+     * @param _n ratio numerator
+     * @param _d ratio denominator
+     * @param _time the sampling timestamp
+     */
+    event SampleAdded(IERC20Token indexed _token, uint256 _n, uint256 _d, uint256 _time);
+
+    constructor() public {
+        // set up administrative roles.
+        _setRoleAdmin(ROLE_OWNER, ROLE_OWNER);
+        _setRoleAdmin(ROLE_SEEDER, ROLE_OWNER);
+
+        // allow the deployer to initially govern the contract.
+        _setupRole(ROLE_OWNER, msg.sender);
+    }
+
+    modifier initialized(IERC20Token _token) {
+        _initialized(_token);
+        _;
+    }
+
+    // error message binary size optimization
+    function _initialized(IERC20Token _token) internal view {
+        require(isInitialized(_token), "ERR_NOT_INITIALIZED");
+    }
+
+    /**
+     * @dev initializes the accumulator for a specific token
+     * can only be called by an owner or a seeder
+     *
+     * @param _token the token the data is accumulated for
+     * @param _startTime the accumulation starting time
+     */
+    function initialize(IERC20Token _token, uint256 _startTime) external override validAddress(address(_token)) {
+        require(hasRole(ROLE_OWNER, msg.sender) || hasRole(ROLE_SEEDER, msg.sender), "ERR_ACCESS_DENIED");
+        require(_startTime <= time(), "ERR_INVALID_TIME");
+
+        // make sure that we're not attempting to initialize the accumulator twice
+        require(!isInitialized(_token), "ERR_ALREADY_INITIALIZED");
+
+        addSample(_token, Fraction({ n: INITIAL_SAMPLE_N, d: INITIAL_SAMPLE_D }), _startTime);
+
+        emit Initialized(_token, _startTime);
+    }
+
+    /**
+     * @dev adds a new sample to the accumulator
+     * can only be called by an owner
+     *
+     * @param _token the token the data is accumulated for
+     * @param _n ratio numerator
+     * @param _d ratio denominator
+     */
+    function addSample(
+        IERC20Token _token,
+        uint256 _n,
+        uint256 _d
+    ) external override validAddress(address(_token)) greaterThanZero(_d) initialized(_token) {
+        require(hasRole(ROLE_OWNER, msg.sender), "ERR_ACCESS_DENIED");
+
+        addSample(_token, Fraction({ n: _n, d: _d }), time());
+    }
+
+    /**
+     * @dev adds a past sample to the accumulator
+     * can only be called by a seeder to add past samples
+     *
+     * @param _token the token the data is accumulated for
+     * @param _n ratio numerator
+     * @param _d ratio denominator
+     * @param _time the sampling timestamp
+     */
+    function addPastSample(
+        IERC20Token _token,
+        uint256 _n,
+        uint256 _d,
+        uint256 _time
+    ) external override validAddress(address(_token)) greaterThanZero(_d) initialized(_token) {
+        require(hasRole(ROLE_SEEDER, msg.sender), "ERR_ACCESS_DENIED");
+        require(_time <= time(), "ERR_INVALID_TIME");
```

Comment:
Minor: If `_time == time()` then it's not a "past sample" (and also - the user can/should call function `addSample` in this case).
But in either case, see my comment above about renaming this function to `addSample` and calling it from the other `addSample`.

Reply:
I'll fix it to a `<`

Reply:
Fixed